### PR TITLE
[BK-1.15][FLINK-26931][Connector/pulsar] Make the producer name and consumer name unique in Pulsar

### PR DESCRIPTION
**Pull request by syhily**
This is a backport PR for #19285.

### CODE REVIEW

**Comment by syhily** [comment]
@flinkbot run azure

**Comment by syhily** [comment]
@AHeise This is a backport PR. Can you review it?

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/config/PulsarClientFactory.java [hunk 1]
```diff
@@ -153,7 +153,7 @@ public static PulsarClient createClient(PulsarConfiguration configuration) {
 
     /**
      * PulsarAdmin shares almost the same configuration with PulsarClient, but we separate this
-     * create method for directly creating it.
+     * creating method for directly use it.
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/config/PulsarClientFactory.java:156]: Nit: I am having a hard time understanding this doc string. Can you maybe rephrase it?

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/config/PulsarSinkConfigUtils.java [hunk 2]
```diff
@@ -70,7 +71,10 @@ public static <T> ProducerBuilder<T> createProducerBuilder(
             PulsarClient client, Schema<T> schema, SinkConfiguration configuration) {
         ProducerBuilder<T> builder = client.newProducer(schema);
 
-        configuration.useOption(PULSAR_PRODUCER_NAME, builder::producerName);
+        configuration.useOption(
```
Comment (fapaul) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/config/PulsarSinkConfigUtils.java:74]: What happens during an application restart? I guess a new producerName is generated. Is this a problem?

**Review by fapaul — COMMENTED** [review]
Since this PR fixes an actual bug, I would expect to add a test case. Can you add one so that the issue does not arise again?

**Comment by AHeise** [comment]
> Since this PR fixes an actual bug, I would expect to add a test case. Can you add one so that the issue does not arise again?

I think this is hard to test on integration level (you probably need to check on Pulsar server if a specific producer is registered) and the change is rather implicitly tested with other tests.
However, if there is an easy way to test, I'm also in favor of adding it.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/config/PulsarClientFactory.java (hunk 1 above)
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/common/config/PulsarClientFactory.java:156]: OK.

**Comment by syhily** [comment]
> Since this PR fixes an actual bug, I would expect to add a test case. Can you add one so that the issue does not arise again?

I think this could be hard to test. Because we just use the same producer name which is not allowed on the Pulsar. We just need to make sure it was unique.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/config/PulsarSinkConfigUtils.java (hunk 2 above)
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/config/PulsarSinkConfigUtils.java:74]: This wouldn't be a problem. Because we promise consistency by using transactions.
